warehouse-management-service/main.go

```diff
@@ -4,45 +4,57 @@ import (
 	"fmt"
 	"net/http"
 	"os"
+	"warehouse-management-service/internal/config"
+	"warehouse-management-service/internal/handler"
+	"warehouse-management-service/pkg/database/postgres"
 
-	"github.com/gin-gonic/gin"
+	_ "github.com/lib/pq"
+)
+
+const (
+	CONFIG_FILE_PATH_ENV_VARIABLE = "CONFIG_FILE_PATH"
 )
 
 func main() {
-	err := initialize()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "App startup error: %v", err)
-		os.Exit(1)
+	// Get absolute path of config file from env variable
+	configFilePath, ok := os.LookupEnv(CONFIG_FILE_PATH_ENV_VARIABLE)
+	if !ok {
+		fmt.Fprintf(os.Stderr, "%s env variable not set", CONFIG_FILE_PATH_ENV_VARIABLE)
 	}
-}
 
-func initialize() error {
-	err := setupDatabase()
+	// Get config
+	config, err := config.FromFile(configFilePath)
 	if err != nil {
-		return err
+		fmt.Fprintf(os.Stderr, "App startup error: %v", err)
+		os.Exit(1)
 	}
 
-	err = startServer()
+	// Get Postgres DB service
+	db, err := postgres.New(config.Postgres.Host,
+		config.Postgres.Port,
+		config.Postgres.Username,
+		config.Postgres.Password,
+		config.Postgres.DBName,
+		config.Postgres.SSLMode)
 	if err != nil {
-		return err
+		fmt.Fprintf(os.Stderr, "App startup error: %v", err)
+		os.Exit(1)
 	}
 
-	return nil
-}
-
-func setupDatabase() error {
-	// TODO
-	return nil
-}
+	// Close DB connection at shutdown
+	defer func() {
+		err = db.Close()
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+	}()
 
-func startServer() error {
-	router := gin.Default()
+	handler := handler.New(db)
 
-	router.GET("/ping", func(ctx *gin.Context) {
-		ctx.JSON(http.StatusOK, gin.H{
-			"message": "pong",
-		})
-	})
+	http.ListenAndServe(":80", handler)
 
-	return router.Run(":80")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "App startup error: %v", err)
+		os.Exit(1)
+	}
```

Comment:
Oh hey. I just realised something else. From the [`ListenAndServe()`](https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/net/http/server.go;drc=76d39ae3499238ac7efb731f4f4cd47b1b3288ab;l=2976) method docs:
```
// ListenAndServe always returns a non-nil error. After Shutdown or Close,
// the returned error is ErrServerClosed.
```

So `ListenAndServe` is guaranteed to return a non-nil error value. So your check after this line should be something along the lines of:

```go
switch err {
case ErrServerClosed:
    log(INFO, "shutting down...")
default:
    log(ERROR, "error starting up server: %v", err)
}
```